### ISSUE TITLE
Consider optimizing folds by not using variadic expansion

### ISSUE THREAD

**Issue by ldionne** [boostorg/hana#268]
Right now, the folds are implemented by unpacking the sequences and then folding recursively on variadic packs (with some unrolling). However, passing many parameters around can be difficult to optimize away, and overload resolution is expansive in terms of compile-time. Instead, we should consider having a unrolled recursive fold that indexes into the original structure.

**Comment by jfalcou** [comment]
We use ... and a creative use of the for_each_args like function in Boost.SIMD to perform runtime fold. Not sure it helps you but it still doesnt require recursion : 

https://github.com/NumScale/boost.simd/blob/develop/include/boost/simd/arch/common/generic/function/autofold.hpp#L47

(extract<N>(a) is basically a[N] fyi)

Slap me if I misunderstood what you wanted here.

**Comment by ldionne** [comment]
Interesting, but IIUC the result type of your fold is completely determined by the result of calling the function `F` on your `Arg::value_type`, which is fixed. In other words, that's not a heterogeneous fold, it's an homogeneous fold, albeit an unrolled one.

**Comment by jfalcou** [comment]
Yes, as I said, i was more showing the trick with the unrolling. I *think* the initialisation can be made heterogeneous though.

**Comment by ldionne** [comment]
Oh, ah okay I see. Yeah there's a way to do it but it requires placement new, which can't be constexpr. Otherwise you have to require that every intermediate type is default constructible and assignable, which is too strict for us.

**Comment by ldionne** [comment]
The idea of what I'm talking about is implemented [here](https://github.com/boostorg/hana/blob/heterogeneous_storage/experimental/heterogeneous_storage_scan_left.hpp#L44). It's a scan, not a fold, but the fold is easy to derive from the scan implementation.

**Comment by pfultz2** [comment]
You couldn't do something like:

```cpp
template<class Index, class Sequence, class State, class Op>
constexpr auto fold_3(Index i, Sequence s, State state, Op op)
{
    auto state1 = op(state, s[i]);
    auto state2 = op(state1, s[i+1]);
    return op(state2, s[i+2]);
}
```

**Comment by ldionne** [comment]
Yeah, that's exactly what I had in mind.

**Comment by ldionne** [comment]
We'd have a few of them unrolled, and then we'd chain unrolled calls until the whole length of the sequence has been processed.
